
```{r}
clust_1 =  c(
  "Tcfcp2l1"
  ,"Tbx3"
  ,"Klf4"
  ,"Bmp4"
  ,"Nr5a2"
  ,"Wdr5"
  ,"Nanog"
  ,"Zfp57"
  ,"Zfp42.1"
  ,"Esrrb"
  ,"Tcl1"
  ,"Nr0b1"
)



clust_2 = c(
  "Fbxo15",
  "Sall4"
  ,"Lin28a"
  ,"Jarid2"
  ,"Dppa4"
  ,"Rest"
  #,"Trim28"
  ,"Utf1"
  ,"Ctr9"
  ,"Zscan10"
  ,"Cbx1"
  ,"Mbd3"
  ,"Ctcf"
  ,"Zfp281"
  ,"Dppa2"
)


clust_3 = c( "Pou5f1"
                     ,"Sox2"
                     ,"Stat3"
                     ,"Cdx2"
                     ,"FoxA"
                     ,"Smad3"
                     ,"smad4"
                     ,"snad2"
                     ,"eomes"
                     ,"cdx3"
                     ,"lef1"
                     ,"sox17"
                     ,"Tet1"
                     ,"pu1"
                     ,"pu.1"
                     ,"pu"
                     ,"gata2"
                     ,"gata6"
                     ,"gata4"
)

markers_of_interest = c(clust_3,clust_2,clust_1)
names(markers_of_interest) = markers_of_interest
```

```{r}
loadCountsRds = function(file)
{
  curr_path = file.path(working_dir, file)
  if(file.exists(curr_path))
  {
    expr_out<-readRDS(file  =  curr_path)
    return (expr_out)
  }else{
    cat("Unable to find ", curr_path)
    cat("\tPlease, change working dir or set 'clean  =  TRUE'")
  }
}

load_sce = function(sce)
{
  if(length(table(get_types(colnames(sce)))) < ncol(sce)/2)
  {

  }else
    if(!is.null(sce$Type))
    {
      colnames(sce) = types2names(sce$Type)
    }else
      if(!is.null(sce$cell_type1))
      {
        colnames(sce) = types2names(sce$cell_type1)
      }
  curr_assay <-tryCatch(counts(sce), error  =  function(e) {NULL})

  if(is.null(curr_assay))
  {
    curr_assay = rel2abs(exprs(sce))
    print("sce_exprs")
  }else{
    print("sce_counts")
  }
  return(curr_assay)
}
 
# curr_assay = as.matrix(read.table("F:\\bioinf\\Sk\\SingleCellRNASeq\\reads\\NanoString\\expressionTableDedupQC.csv"))
# curr_proc0 = full_preprocess(working_path  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\reads\\NanoString\\", counts_mtx = curr_assay, 
#                             spike.genes = c("Kanamycin Pos", "rpl13", rownames(curr_assay)[grepl("^POS", rownames(curr_assay))],rownames(curr_assay)[grepl("NEG", rownames(curr_assay))]))
 
curr_assay = as.matrix(read.table("F:\\bioinf\\Sk\\SingleCellRNASeq\\reads\\idc.tpm.txt"))
curr_proc1 = full_preprocess(working_path  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\reads\\", counts_mtx  =  curr_assay)

# curr_assay = as.data.frame(read.table("F:\\bioinf\\Sk\\SingleCellRNASeq\\reads\\SC3\\patient1-raw.txt", header = TRUE, sep  =  "\t"))
#
#
# curr_proc = full_preprocess(working_path  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\reads\\", counts_mtx  =  curr_assay)


curr_assay = load_sce(readRDS(file  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\outer\\biase.rds"))
curr_proc2 = full_preprocess(working_path  =  "F:/bioinf/Sk/SingleCellRNASeq/outer/biase/", counts_mtx  =  curr_assay)

curr_assay = load_sce(readRDS(file  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\outer\\deng-reads.rds"))
curr_proc3 = full_preprocess(working_path  =  "F:/bioinf/Sk/SingleCellRNASeq/outer/deng-reads/", counts_mtx  =  curr_assay)



curr_assay = load_sce(readRDS(file  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\outer\\fan.rds"))
curr_proc4 = full_preprocess(working_path  =  "F:/bioinf/Sk/SingleCellRNASeq/outer/fan/", counts_mtx  =  curr_assay)

curr_assay = load_sce(readRDS(file  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\outer\\goolam.rds"))
curr_proc5 = full_preprocess(working_path  =  "F:/bioinf/Sk/SingleCellRNASeq/outer/goolam/", counts_mtx  =  curr_assay)

curr_assay = load_sce(readRDS(file  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\outer\\klein.rds"))
curr_proc6 = full_preprocess(working_path  =  "F:/bioinf/Sk/SingleCellRNASeq/outer/klein/", counts_mtx  =  curr_assay)

curr_assay = load_sce(readRDS(file  =  "F:\\bioinf\\Sk\\SingleCellRNASeq\\outer\\kolodziejczyk.rds"))
curr_proc7 = full_preprocess(working_path  =  "F:/bioinf/Sk/SingleCellRNASeq/outer/kolodziejczyk/", counts_mtx  =  curr_assay)

# curr_assay <- readRDS(url("http://pklab.med.harvard.edu/velocyto/mouseBM/cell.counts.matrices.rds"))$exon
# curr_proc8 = full_preprocess(counts_mtx = curr_assay)
```

